src/Builders/PromptBuilder.php

```diff
@@ -1040,15 +1090,18 @@ private function getConfiguredModel(CapabilityEnum $capability): ModelInterface
     {
         $requirements = ModelRequirements::fromPromptData($capability, $this->messages, $this->modelConfig);
 
-        // If a model has been explicitly set, return it
         if ($this->model !== null) {
+            // Explicit model was provided via usingModel(); just update config and bind dependencies.
             $this->model->setConfig($this->modelConfig);
             $this->registry->bindModelDependencies($this->model);
             return $this->model;
         }
 
-        // Find a suitable model based on requirements
+        /** @var list<array{0:string,1:ModelMetadata}> $candidateModels */
+        $candidateModels = [];
```

Comment:
This is not very efficient. The model lists can be massive, so iterating over them for every model preference is slow. I'd suggest to use a map with keys as follows:
* The provider ID and model ID concatenated.
* The model ID on its own.

We can also include some sort of identifier (e.g. `providerModel::{provider ID}::{model ID}`) and `model::{model ID}`) to make sure there cannot be any conflicts.

This way it's super quick to look up the correct provider ID and `ModelMetadata`.

Reply:
Good idea! Implemented in 722e4aa